pdns/recursordist/rec-rust-lib/rust/src/web.rs

```diff
@@ -0,0 +1,1143 @@
+            cxx::let_cxx_string!(s = &kv.value);
+            if kv.key == "x-api-key" && ctx.api_ch.as_ref().unwrap().matches(&s) {
+                auth_ok = true;
+                break;
+            }
+        }
+    }
+    if !auth_ok && allow_password {
+        auth_ok = compare_authorization(ctx, reqheaders);
+        if !auth_ok {
+            rustmisc::log(
+                logger,
+                rustweb::Priority::Debug,
+                "Authentication failed",
+                &vec![rustmisc::KeyValue {
+                    key: "urlpath".to_string(),
+                    value: request.uri.to_owned(),
+                }],
+            );
+            unauthorized(response, headers, "Basic");
+            return;
+        }
+    }
+    if !auth_ok {
+        rustmisc::log(
+            logger,
+            rustweb::Priority::Error,
+            "Authentication failed",
+            &vec![rustmisc::KeyValue {
+                key: "urlpath".to_string(),
+                value: request.uri.to_owned(),
+            }],
+        );
+        unauthorized(response, headers, "X-API-Key");
+        return;
+    }
+    response.status = StatusCode::OK.as_u16(); // 200;
+
+    headers.insert(
+        header::X_CONTENT_TYPE_OPTIONS,
+        header::HeaderValue::from_static("nosniff"),
+    );
+    headers.insert(
+        header::X_FRAME_OPTIONS,
+        header::HeaderValue::from_static("deny"),
+    );
+    headers.insert(
+        header::HeaderName::from_static("x-permitted-cross-domain-policies"),
+        header::HeaderValue::from_static("none"),
+    );
+    headers.insert(
+        header::X_XSS_PROTECTION,
+        header::HeaderValue::from_static("1; mode=block"),
+    );
+    headers.insert(
+        header::CONTENT_SECURITY_POLICY,
+        header::HeaderValue::from_static("default-src 'self'; style-src 'self' 'unsafe-inline'"),
+    );
+
+    // This calls into C++
+    match handler(request, response) {
+        Ok(_) => {}
+        Err(_) => {
+            let status = StatusCode::UNPROCESSABLE_ENTITY; // 422
+            response.status = status.as_u16();
+            response.body = status.canonical_reason().unwrap().as_bytes().to_vec();
+        }
+    }
+}
+
+// Data used by requests handlers, only counter is r/w.
+struct Context {
+    urls: Vec<String>,
+    password_ch: cxx::UniquePtr<rustweb::CredentialsHolder>,
+    api_ch: cxx::UniquePtr<rustweb::CredentialsHolder>,
+    acl: cxx::UniquePtr<rustmisc::NetmaskGroup>,
+    logger: cxx::SharedPtr<rustmisc::Logger>,
+    loglevel: rustmisc::LogLevel,
+}
+
+// Serve a file
+fn file(
+    ctx: &Context,
+    method: &Method,
+    path: &str,
+    request: &rustweb::Request,
+    response: &mut rustweb::Response,
+) {
+    let mut uripath = path;
+    if uripath == "/" {
+        uripath = "/index.html";
+    }
+    let pos = ctx
+        .urls
+        .iter()
+        .position(|x| String::from("/") + x == uripath);
+    if pos.is_none() {
```

Comment:
It's not clear to me why we are doing this lookup?

Reply:
Originally I had a plan to fully implement the file access in Rust, but that plan changed. Some remains are left, will remove.